pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -133,21 +163,41 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(cb *ClusterBuilder,
 				defaultCluster.cluster.AltStatName = util.BuildStatPrefix(cb.push.Mesh.OutboundClusterStatName, string(service.Hostname), "", port, service.Attributes)
 			}
 
-			subsetClusters := cb.applyDestinationRule(defaultCluster, DefaultClusterMode, service, port, networkView)
+			subsetClusters := cb.applyDestinationRule(defaultCluster, DefaultClusterMode, service, port, networkView, clusterKey.destinationRule)
 
-			clusters = cp.conditionallyAppend(clusters, nil, defaultCluster.build())
-			clusters = cp.conditionallyAppend(clusters, nil, subsetClusters...)
+			if patched := cp.apply(nil, defaultCluster.build()); patched != nil {
+				cc := &discovery.Resource{Name: patched.Name, Resource: util.MessageToAny(patched)}
+				resources = append(resources, cc)
+				cb.cache.Add(clusterKey, tokens[clusterKey.clusterName], cc)
+			}
+			for _, ss := range subsetClusters {
+				if patched := cp.apply(nil, ss); patched != nil {
+					nk := *clusterKey
+					nk.clusterName = ss.Name // TODO this is not sufficient, other keys may change as well. At least envoyfilters
+					cc := &discovery.Resource{Name: patched.Name, Resource: util.MessageToAny(patched)}
+					resources = append(resources, cc)
+					cb.cache.Add(&nk, tokens[ss.Name], cc)
+				}
+			}
 		}
 	}
+	log.Errorf("howardjohn: hits=%v miss=%v", hit, miss)
 
-	return clusters
+	return resources
 }
 
 type clusterPatcher struct {
 	efw  *model.EnvoyFilterWrapper
 	pctx networking.EnvoyFilter_PatchContext
 }
 
+func (p clusterPatcher) apply(hosts []host.Name, c *cluster.Cluster) *cluster.Cluster {
```

Comment:
Where were we doing this before? If we weren't why do we need it for caching?

Reply:
We need to apply the patches on a per-cluster basis now, where-as the old way (conditionallyAppend) did many. I also thing there may be some merge conflict here at some point, though so I probably need to cxlean it up

Reply:
In that case can we dry up conditionallyAppend by calling this there?